File: Cpp/451.sort-characters-by-frequency.cpp
/*
 * @lc app=leetcode id=451 lang=cpp
 *
 * [451] Sort Characters By Frequency
 */

// @lc code=start
class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> map;
        priority_queue<pair<int, char>> pq;
        for (char c : s) {
            map[c]++;
        }
        for (auto it : map) {
            pq.push({it.second, it.first});
        }
        string res = "";
        while (!pq.empty()) {
            res += string(pq.top().first, pq.top().second);
            pq.pop();
        }
        return res;
    }
};
// @lc code=end

File: Go/451.sort-characters-by-frequency.go
/*
 * @lc app=leetcode id=451 lang=golang
 *
 * [451] Sort Characters By Frequency
 */

// @lc code=start
func frequencySort(s string) string {
    freq := make(map[rune]int)
    maxFreq := 0
    for _, ch := range s {
        freq[ch]++
        if freq[ch] > maxFreq {
            maxFreq = freq[ch]
        }
    }
    freqMap := make(map[int][]rune)
    // k: ch, v: freq
    for k, v := range freq {
        freqMap[v] = append(freqMap[v], k)
    }
    var res string
    for maxFreq > 0 {
        if chars, ok := freqMap[maxFreq]; ok {
            for _, ch := range chars {
                for i := maxFreq - 1; i >= 0; i--  {
                    res += string(ch)
                }
            }
        }
        maxFreq--
    }
    return res
}
// @lc code=end

